FileValidator TS Error: Type 'TFile | undefined' is not assignable to type 'File'.

### Did you read the migration guide?

- [X] I have read the whole migration guide

### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Potential Commit/PR that introduced the regression

_No response_

### NestJS version

9.4.3 -> 10.0.5

### Describe the regression

I have a file validator class that inherits from the FileValidator abstract class.
As mentioned in the docs, I typed the file parameter with multer's File interface:

> HINT
> The FileValidator interfaces supports async validation via its isValid function. To leverage type security, you can also type the file parameter as Express.Multer.File in case you are using express (default) as a driver.

but i get this typescript error:

```
src/image.validator.ts:10:9 - error TS2416: Property 'isValid' in type 'ImageFileValidator' is not assignable to the same property in base type 'FileValidator<ImageFileValidatorOptions>'.
  Type '(file: File) => Promise<boolean>' is not assignable to type '<TFile extends IFile = any>(file?: TFile | undefined) => boolean | Promise<boolean>'.
    Types of parameters 'file' and 'file' are incompatible.
      Type 'TFile | undefined' is not assignable to type 'File'.
        Type 'undefined' is not assignable to type 'File'.

10   async isValid(file: Express.Multer.File) {
           ~~~~~~~
```

### Minimum reproduction code

https://stackblitz.com/edit/nestjs-repro-filevalidator

### Input code

```ts
import { FileValidator } from '@nestjs/common';
import sharp, { Metadata, Sharp } from 'sharp';

export interface ImageFileValidatorOptions {
  formats: Metadata['format'][];
  message?: string | ((format: Metadata['format'][]) => string);
}

export class ImageFileValidator extends FileValidator<ImageFileValidatorOptions> {
  async isValid(file: Express.Multer.File) {
    let image: Sharp;
    let metadata: Metadata;

    try {
      image = sharp(file.path ? file.path : file.buffer);
      metadata = await image.metadata();
    } catch (error) {
      return false;
    }

    const { formats } = this.validationOptions;
    return formats.includes(metadata.format);
  }

  buildErrorMessage(file: Express.Multer.File) {
    const { message, formats } = this.validationOptions;

    if (message) {
      if (typeof message === 'function') return message(formats);
      return message;
    }

    return (
      'Image validation failed (expected format to be one of ' +
      formats.join(', ')
    );
  }
}
```


### Expected behavior

No Typescript error.

### Other

_No response_

--- a/packages/common/pipes/file/file-type.validator.ts
+++ b/packages/common/pipes/file/file-type.validator.ts
@@ -16,12 +16,15 @@ export type FileTypeValidatorOptions = {
  *
  * @publicApi
  */
-export class FileTypeValidator extends FileValidator<FileTypeValidatorOptions> {
+export class FileTypeValidator extends FileValidator<
+  FileTypeValidatorOptions,
+  IFile
+> {
   buildErrorMessage(): string {
     return `Validation failed (expected type is ${this.validationOptions.fileType})`;
   }
 
-  isValid<TFile extends IFile = any>(file?: TFile): boolean {
+  isValid(file?: IFile): boolean {
     if (!this.validationOptions) {
       return true;
     }

--- a/packages/common/pipes/file/file-validator.interface.ts
+++ b/packages/common/pipes/file/file-validator.interface.ts
@@ -6,16 +6,17 @@ import { IFile } from './interfaces';
  * @see {ParseFilePipe}
  * @publicApi
  */
-export abstract class FileValidator<TValidationOptions = Record<string, any>> {
+export abstract class FileValidator<
+  TValidationOptions = Record<string, any>,
+  TFile extends IFile = IFile,
+> {
   constructor(protected readonly validationOptions: TValidationOptions) {}
 
   /**
    * Indicates if this file should be considered valid, according to the options passed in the constructor.
    * @param file the file from the request object
    */
-  abstract isValid<TFile extends IFile = any>(
-    file?: TFile,
-  ): boolean | Promise<boolean>;
+  abstract isValid(file?: TFile): boolean | Promise<boolean>;
 
   /**
    * Builds an error message in case the validation fails.

--- a/packages/common/pipes/file/max-file-size.validator.ts
+++ b/packages/common/pipes/file/max-file-size.validator.ts
@@ -13,7 +13,10 @@ export type MaxFileSizeValidatorOptions = {
  *
  * @publicApi
  */
-export class MaxFileSizeValidator extends FileValidator<MaxFileSizeValidatorOptions> {
+export class MaxFileSizeValidator extends FileValidator<
+  MaxFileSizeValidatorOptions,
+  IFile
+> {
   buildErrorMessage(): string {
     if ('message' in this.validationOptions) {
       if (typeof this.validationOptions.message === 'function') {
@@ -26,7 +29,7 @@ export class MaxFileSizeValidator extends FileValidator<MaxFileSizeValidatorOpti
     return `Validation failed (expected size is less than ${this.validationOptions.maxSize})`;
   }
 
-  public isValid<TFile extends IFile = any>(file?: TFile): boolean {
+  public isValid(file?: IFile): boolean {
     if (!this.validationOptions || !file) {
       return true;
     }

